src/sio.ts

```diff
@@ -33,11 +33,234 @@ const DIV_QUOTIENT = 0x070; //  Divider result quotient
+    this.update();
+  }
+
+  update() {
+    const s32 = BitUtil.s32;
+    const u32 = BitUtil.u32;
+
+    const N = this.index;
+    const ctrl0 = new InterpolatorConfig(this.ctrl0);
+    const ctrl1 = new InterpolatorConfig(this.ctrl1);
+
+    const do_clamp = ctrl0.clamp && N == 1;
+    const do_blend = ctrl0.blend && N == 0;
+
+    ctrl0.clamp = do_clamp;
+    ctrl0.blend = do_blend;
+    ctrl0._reserved0 = 0;
+    ctrl1.clamp = false;
+    ctrl1.blend = false;
+    ctrl1.overf0 = false;
+    ctrl1.overf1 = false;
+    ctrl1.overf = false;
+    ctrl1._reserved0 = 0;
+
+    const input0 = s32(ctrl0.cross_input ? this.accum1 : this.accum0);
+    const input1 = s32(ctrl1.cross_input ? this.accum0 : this.accum1);
+
+    const msbmask0 = ctrl0.mask_msb == 31 ? 0xffffffff : (1 << (ctrl0.mask_msb + 1)) - 1;
+    const msbmask1 = ctrl1.mask_msb == 31 ? 0xffffffff : (1 << (ctrl1.mask_msb + 1)) - 1;
+    const mask0 = msbmask0 & ~((1 << ctrl0.mask_lsb) - 1);
+    const mask1 = msbmask1 & ~((1 << ctrl1.mask_lsb) - 1);
+
+    const uresult0 = (input0 >>> ctrl0.shift) & mask0;
+    const uresult1 = (input1 >>> ctrl1.shift) & mask1;
+
+    const overf0 = Boolean((input0 >>> ctrl0.shift) & ~msbmask0);
+    const overf1 = Boolean((input1 >>> ctrl1.shift) & ~msbmask1);
+    const overf = overf0 || overf1;
+
+    const sextmask0 = (uresult0 & (1 << ctrl0.mask_msb)) ? (-1 << ctrl0.mask_msb) : 0;
+    const sextmask1 = (uresult1 & (1 << ctrl1.mask_msb)) ? (-1 << ctrl1.mask_msb) : 0;
+
+    const sresult0 = uresult0 | sextmask0;
+    const sresult1 = uresult1 | sextmask1;
+
+    const result0 = ctrl0.is_signed ? sresult0 : uresult0;
+    const result1 = ctrl1.is_signed ? sresult1 : uresult1;
+
+    const addresult0 = this.base0 + (ctrl0.add_raw ? input0 : result0);
+    const addresult1 = this.base1 + (ctrl1.add_raw ? input1 : result1);
+    const addresult2 = this.base2 + result0 + (do_blend ? 0 : result1);
+
+    const uclamp0 = u32(result0) < u32(this.base0) ? this.base0 : (u32(result0) > u32(this.base1) ? this.base1 : result0);
+    const sclamp0 = s32(result0) < s32(this.base0) ? this.base0 : (s32(result0) > s32(this.base1) ? this.base1 : result0);
+    const clamp0 = ctrl0.is_signed ? sclamp0 : uclamp0;
+
+    const alpha1 = result1 & 0xff;
+    const ublend1 = u32(this.base0) + (((alpha1 * (u32(this.base1) - u32(this.base0))) / 256) | 0);
+    const sblend1 = s32(this.base0) + (((alpha1 * (s32(this.base1) - s32(this.base0))) / 256) | 0);
+    const blend1 = ctrl1.is_signed ? sblend1 : ublend1;
+
+    this.smresult0 = u32(result0);
+    this.smresult1 = u32(result1);
+    this.result0 = u32(do_blend ? alpha1 : (do_clamp ? clamp0 : addresult0) | (ctrl0.force_msb << 28));
+    this.result1 = u32((do_blend ? blend1 : addresult1) | (ctrl0.force_msb << 28));
+    this.result2 = u32(addresult2);
+
+    ctrl0.overf0 = overf0;
+    ctrl0.overf1 = overf1;
+    ctrl0.overf = overf;
+    this.ctrl0 = u32(ctrl0.toUint32());
+    this.ctrl1 = u32(ctrl1.toUint32());
```

Comment:
It makes sense to move the `u32` call into `toUint32()`, since the function name implies that it returns an unsigned int.

Reply:
oh yes, that definitely makes sense

Reply:
I noticed the u32 is not necessary at all.
due to the bit operations in InterpCtrl, all resulting values are positive anyway (bit 31 is part of the reserved bits), meaning the u32 call is redundant

Reply:
done